src/Workspaces/CSharp/Portable/CodeGeneration/CSharpCodeGenerationHelpers.cs

```diff
@@ -21,6 +20,67 @@ namespace Microsoft.CodeAnalysis.CSharp.CodeGeneration
 {
     internal static class CSharpCodeGenerationHelpers
     {
+        public static void AdjustNullableAnnotationByAttributes(ITypeSymbol type, ref ImmutableArray<AttributeData> attributes, ref NullableAnnotation nullableAnnotation, bool isParameter)
+        {
+            if (type.IsReferenceType)
+            {
+                if (!(type is ITypeParameterSymbol typeParameter) ||
+                    typeParameter.ReferenceTypeConstraintNullableAnnotation == NullableAnnotation.NotAnnotated)
+                {
+                    // For [AllowNull]/[MaybeNull], add the nullable annotation.
+                    // ([AllowNull] string?) -> (string?)
+                    // ([AllowNull] string)  -> (string?)
+                    var newAttributes = attributes.RemoveAll(IsAllowNullOrMaybeNullAttribute);
+                    if (newAttributes.Length != attributes.Length)
+                    {
+                        attributes = newAttributes;
+                        if (nullableAnnotation == NullableAnnotation.NotAnnotated)
+                        {
+                            nullableAnnotation = NullableAnnotation.Annotated;
+                        }
+                    }
+
+                    if (isParameter)
+                    {
+                        // The compiler does not allow parameter nullabilities to differ.
+                        // - ([DisallowNull] string?) -> (string)
+                        // + ([DisallowNull] string?) -> ([DisallowNull] string?)
+                        // ([DisallowNull] string)  -> (string)
+                        if (nullableAnnotation == NullableAnnotation.NotAnnotated)
+                        {
+                            attributes = attributes.RemoveAll(IsDisallowNullOrNotNullAttribute);
+                        }
+                    }
+                    else
+                    {
+                        // For [DisallowNull]/[NotNull], remove the nullable annotation.
+                        // ([DisallowNull] string?) -> (string)
+                        // ([DisallowNull] string)  -> (string)
+                        newAttributes = attributes.RemoveAll(IsDisallowNullOrNotNullAttribute);
+                        if (newAttributes.Length != attributes.Length)
+                        {
+                            attributes = newAttributes;
+                            if (nullableAnnotation == NullableAnnotation.Annotated)
+                            {
+                                nullableAnnotation = NullableAnnotation.NotAnnotated;
+                            }
+                        }
+                    }
+                }
+            }
+            else if (type.IsValueType)
```

Comment:
does this also handle type parameters with the `struct` constraint?

Reply:
Yes.

Reply:
Yes. There are two tests for this:

https://github.com/dotnet/roslyn/pull/38203/files/7804c685e5c13a18d0fb7ab8030565a11d920cc7#diff-a5084bcce45367aadfd748c52b57f140R1911
https://github.com/dotnet/roslyn/pull/38203/files/7804c685e5c13a18d0fb7ab8030565a11d920cc7#diff-a5084bcce45367aadfd748c52b57f140R1950